### PR TITLE
Sketch out profile manager.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: SignalServiceKit/src/Profiles/OWSProfilesManager.m [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+#pragma mark - Local Profile Key
+
++ (NSData *)generateLocalProfileKey
+{
+    // TODO:
+    OWSFail(@"Profile key generation is not yet implemented.");
+    return [SecurityUtils generateRandomBytes:kProfileKeyLength];
+}
+
+- (nullable NSData *)localProfileKey
```
Comment (michaelkirk) [inline SignalServiceKit/src/Profiles/OWSProfilesManager.m]: Is this really nullable? If so, our assert is going to fail.

File: SignalServiceKit/src/Profiles/OWSProfilesManager.m [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    _messageSender = messageSender;
+
+    OWSSingletonAssert();
+
+    // Register this manager with the message sender.
+    // This is a circular dependency.
+    [messageSender setProfilesManager:self];
+
+    // Try to load.
+    _localProfileKey = [self.storageManager objectForKey:kOWSProfilesManager_LocalProfileKey
```
Comment (michaelkirk) [inline SignalServiceKit/src/Profiles/OWSProfilesManager.m]: As a pattern, I'm starting to resent using the `storageManager`'s internal `dbConnection` for anything. 

Instead, I think I prefer the general approach that each "manager"-like-thing has it's own dbConnection, which we use to read/write any of the objects "owned" by that manager. 

WDYT @charlesmchen?

File: SignalServiceKit/src/Profiles/OWSProfilesManager.m (hunk 1 above)
Comment (charlesmchen) [inline SignalServiceKit/src/Profiles/OWSProfilesManager.m]: Yup, this is just a first draft of a first draft.  It's already been slightly tweaked in the next PR.

File: SignalServiceKit/src/Profiles/OWSProfilesManager.m (hunk 2 above)
Comment (charlesmchen) [inline SignalServiceKit/src/Profiles/OWSProfilesManager.m]: The key-value accessor methods in TSStorageManager are really useful.  I propose that we move them to a category on `YapDatabaseReadTransaction` and `YapDatabaseReadWriteTransaction`.  

I'm not sure that I support eliminating `TSStorageManager.dbReadConnection` and `TSStorageManager.dbReadWriteConnection` entirely, without reviewing all of the places that they're used. 

I'm making cards "below the line"?

Comment (michaelkirk) [inline SignalServiceKit/src/Profiles/OWSProfilesManager.m]: To clarify, are you proposing leaving this as is and making a card to maybe remove dbReadConnection/dbReadWriteConnection entirely?

Comment (charlesmchen) [inline SignalServiceKit/src/Profiles/OWSProfilesManager.m]: Yup.

* https://trello.com/c/urYDlexK/1264-consider-eliminating-tsstoragemanagerdbreadconnection-and-tsstoragemanagerdbreadwriteconnection
* https://trello.com/c/ndz0SEY7/1263-move-key-value-accessor-methods-from-tsstoragemanager-to-a-category-on-yapdatabasereadtransaction-and-yapdatabasereadwritetransa

**Pull request merged**